### PR TITLE
WIP feat(auth): allow backend server access to our API with JWT authentication

### PR DESCRIPTION
**Pull request by rfk**
@dcoates here's my sketch-in-progress of how we might do backend service auth using JWTs.  Except it doesn't do any of the actual JWT stuff yet...

The idea is is to add a new auth strategy "backendJWT" that accepts JWTs from trusted sources in some to-be-defined request-signing format, and then use it as another auth option on existing endpoints that we have for querying account data.

A service like basket, upon receiving an account-verified notification, could turn around and make an authenticated backend request to `/account/status?uid=blah` to retrieve the email and locale information in a secure fashion.

Assuming we can model security/permissions appropriate in such a system, what do you think of shaping things this way?  /cc @ckarlof as we've talked about this a few times in the past.

### CODE REVIEW

**Comment by rfk** [comment]
OK, I fleshed this out with an actual JWT-based auth scheme, inspirted by Hawk and the way Atlassian does it with JWTs in their web apis.  Definitely not ready for merge, but at least it's something concrete we can discuss.

**Comment by ckarlof** [comment]
> inspirted by Hawk and the way Atlassian does it with JWTs in their web apis

I like at least one of those two things. :)

File: server/hawt-auth.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+      return {
+
+        authenticate: function (request, reply) {
+          var authz = request.headers.authorization
+          if (!authz || authz.indexOf('Hawt ') !== 0) {
+            // AFAICT, we must return an instance of boom.unauthorized here
+            // to indicate "not our auth scheme" and allow other stragies to
+            // handle the auth.  Returning some other error causes it to
+            // fail out with a 401.
+            return reply(boom.unauthorized(null, 'Hawt'))
```
Comment (dannycoates) [inline server/hawt-auth.js]: yes. Its cryptic but it is the documented way to "pass"

> If the err does not include a message but does include a scheme name (e.g. Boom.unauthorized(null, 'Custom')), additional strategies will be attempted in order of preference.

https://github.com/hapijs/hapi/blob/master/API.md#serverauthschemename-scheme

File: server/hawt-auth.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                if (!claims.aud || claims.aud !== config.domain) {
+                  return d.reject('TODO: a new errno?')
+                }
+                if (!claims.nce) {
+                  return d.reject(error.invalidNonce())
+                }
+                if (!claims.qsh) {
+                  return d.reject(error.invalidSignature())
+                }
+                claims.jku = decoded.header.jku
```
Comment (dannycoates) [inline server/hawt-auth.js]: maybe jwtool could take an optional `joi` validation schema on `verify`?

File: server/hawt-auth.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+ *  The JWT claims must include the following meta-data about the request in
+ *  order to guard against reply etc:
+ *
+ *    - iss:  hostname of the service issuing the request
+ *    - aud:  must match publicly-visible domain of the target server
+ *    - exp:  must be a timestamp within configurable freshness window
+ *    - iat:  if present, must be a timestamp within configurable window
+ *    - nce:  nonce; a random string unique to this request
+ *    - qsh:  query hash; a digest of the request parameters (see below)
+ *    - psh:  payload hash; an optional hash of the request payload (see below)
```
Comment (dannycoates) [inline server/hawt-auth.js]: hash all the things!

File: server/hawt-auth.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                  resource: request.url,
+                  nonce: claims.nce,
+                  exp: claims.exp,
+                },
+              }
+            }
+          )
+          .done(
+            function (result) {
+              return reply.continue(result)
```
Comment (dannycoates) [inline server/hawt-auth.js]: nit: could bump this up a `then`

File: server/hawt-auth.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+          }
+          throw new Error('TODO: payload verification not implemented yet')
+        }
+
+      }
+    })
+    next()
+  }
+
+  function decodeAndVerifyJWT(token) {
```
Comment (dannycoates) [inline server/hawt-auth.js]: hopefully jwtool can make this fn cleaner

**Comment by dannycoates** [comment]
I like the direction! I humbly opt out of the security side of the discussion though.

**Comment by rfk** [comment]
> hopefully jwtool can make this fn cleaner

Yep, keen to rebase this atop of #902

**Comment by rfk** [comment]
> > inspirted by Hawk and the way Atlassian does it with JWTs in their web apis
> I like at least one of those two things. :)

I have prepared a defense of the other!  (Although not of my spelling - "inspirted"?  Sounds unpleasant...)

Briefly:

* The timestamp trouble we had with Hawk, we already have with JWTs via `exp` and `iat`.
* The Atlassian scheme doesn't use a nonce AFAICT.  It doesn't seem complex to add one in for extra replay protection, although I could be easily convinced that we're never going to check it on the server and hence should rip it out.
* Hawk's hashing scheme is much simpler than Atlassian's, which is more reminiscent of old-school three-legged-OAuth with its normalization of query params etc.  I don't want that kind of complexity in request normalization.

**Comment by Natim** [comment]
How would the credentials shared between services ? Will it be configuration ?

**Comment by ckarlof** [comment]
I have an alternative idea. Add or extend an endpoint on the OAuth server to allow a JWT for OAuth token exchange, similar to the BrowserID assertion for OAuth token exchange. This would allow any of our services to access our OAuth service providers without modifying the service provider endpoints. It doesn't really solve the Auth server modification issue, but we're going to need a change there regardless.

**Comment by rfk** [comment]
> Add or extend an endpoint on the OAuth server to allow a JWT for OAuth token exchange,
> similar to the BrowserID assertion for OAuth token exchange

A token scoped to a particular uid, or a token tied only to the service, or perhaps the option of both?

I like the idea of handling much of this transparently in the oauth server.  I'm trying to remember why we shied away from doing "oauth backend service tokens" in previous discussions.

I wonder if we'd even need the JWT in this case, or whether we could just give everyone a client_id and client_secret and have them use those directly to auth, in parallel to how it's already done for the token exchange.

**Comment by ckarlof** [comment]
> I wonder if we'd even need the JWT in this case, or whether we could just give everyone a client_id and client_secret and have them use those directly to auth, in parallel to how it's already done for the token exchange.

That would be a simple approach, but has obvious less desirable security properties. I don't mind giving internal backend services a client_id, but giving them a secret seems like it'll create potential confusion and footguns. I'm thinking about bugs where an external client could request any token for a user just using its client_secret because we screwed up some permission check. Having these kind of token requests on a different code path seems to make these kinds of errors less likely.

**Comment by rfk** [comment]
>> A token scoped to a particular uid, or a token tied only to the service, or perhaps the option of both?
> an external client could request any token for a user

OK, so this includes "a token scoped to a particular uid".  I'm cautiously positive about this idea.  /cc @seanmonstar

**Comment by ametaireau** [comment]
> I wonder if we'd even need the JWT in this case, or whether we could just give everyone a client_id and client_secret and have them use those directly to auth, in parallel to how it's already done for the token exchange.

That's the option I was envisioning as well, and has the value of not requiring yet another authentication scheme from the services which want to integrate with the ecosystem.

Also, if we don't chose to do OAuth for service to service auth, I would be interested about why we would like to go away from hawk, given the fact we already have a bunch of libraries for handling authentication with it.

**Comment by Natim** [comment]
Also I like the fact that hawk prevents replay attacks and protect user credentials (client_secret in that case) We could directly use client_id as the ``hawk id`` and client_secret as the ``hawk key``

**Comment by rfk** [comment]
OK then, let's migrate the oauth-service-token discussion into this existing oauth-service-token bug: https://github.com/mozilla/fxa-oauth-server/issues/125

**Comment by rfk** [comment]
For the purposes of this repo, we can have the `/account/status` endpoint just accept an ordinary "profile"-scoped oauth token for reading the profile data.  I don't see much value in inventing a separate scope for this particular endpoint.

It still feels kinda icky to have the auth-server accepting oauth tokens, but watevz, I can't pinpoint anything *wrong* with doing it that way so we might as well take the path of least resistance.

**Comment by rfk** [comment]
> For the purposes of this repo, we can have the /account/status endpoint just accept an ordinary "profile"-scoped oauth token for reading the profile data.

See #915

**Pull request closed without merging**